cpp/src/community/detail/common_methods.cuh

```diff
@@ -183,10 +187,67 @@ weight_t compute_modularity(
   return Q;
 }
 
+template <typename map_key_t, typename map_value_t, bool multi_gpu>
+rmm::device_uvector<map_value_t>&& lookup_primitive_values_for_keys(
+  raft::handle_t const& handle,
+  rmm::device_uvector<map_key_t>& map_keys,
+  rmm::device_uvector<map_value_t>& map_values,
+  rmm::device_uvector<map_key_t>& keys_to_lookup)
+{
+  static_assert(std::is_integral<map_key_t>::value);
+  static_assert(std::is_same<map_key_t, std::int32_t>::value ||
+                std::is_same<map_key_t, std::int64_t>::value);
+
+  static_assert(std::is_arithmetic<map_value_t>::value);
+
+  rmm::device_uvector<map_value_t> values_for_sought_keys(0, handle.get_stream());
+
+  if constexpr (multi_gpu) {
+    cugraph::detail::compute_gpu_id_from_ext_vertex_t<map_key_t> vertex_to_gpu_id_op{
+      handle.get_comms().get_size()};
+
+    kv_store_t<map_key_t, map_value_t, false> cluster_key_weight_map(
+      map_keys.begin(),
+      map_keys.end(),
+      map_values.data(),
+      invalid_vertex_id<map_key_t>::value,
+      std::is_floating_point<map_value_t>::value ? std::numeric_limits<map_value_t>::max()
+                                                 : invalid_vertex_id<map_key_t>::value,
+      handle.get_stream());
+    values_for_sought_keys = cugraph::collect_values_for_keys(handle.get_comms(),
+                                                              cluster_key_weight_map.view(),
+                                                              keys_to_lookup.begin(),
+                                                              keys_to_lookup.end(),
+                                                              vertex_to_gpu_id_op,
+                                                              handle.get_stream());
+  } else {
+    // sort so we can use lower_bound in the transform function
+    thrust::sort_by_key(
+      handle.get_thrust_policy(), map_keys.begin(), map_keys.end(), map_values.begin());
+
+    // for each vertex, look up the vertex weight of the current cluster it is assigned to
+    values_for_sought_keys.resize(keys_to_lookup.size(), handle.get_stream());
+    thrust::transform(handle.get_thrust_policy(),
+                      keys_to_lookup.begin(),
+                      keys_to_lookup.end(),
+                      values_for_sought_keys.begin(),
+                      [d_cluster_weights = map_values.data(),
+                       d_cluster_keys    = map_keys.data(),
+                       num_clusters      = map_keys.size()] __device__(map_key_t cluster) {
+                        auto pos = thrust::lower_bound(
+                          thrust::seq, d_cluster_keys, d_cluster_keys + num_clusters, cluster);
+                        return d_cluster_weights[pos - d_cluster_keys];
```

Comment:
assert(pos != d_cluster_keys + num_clusters);  <= this is executed only in the debug mode, this can happen if `keys_to_look_up` has keys missing in `map_keys`.
thrust::distance(d_cluster_keys, pos);